### ISSUE TITLE
documentation errors in BatSplay.Map.remove

### ISSUE THREAD

**Issue by UnixJunkie** [ocaml-batteries-team/batteries-included#1119]
```
Failure: All:1:(P)Map:2:usual tests on Map.Make:5:test_remove

remove 12 [3,4; 5,6] == [3,4; 5,6]
------------------------------------------------------------------------------
==============================================================================
Failure: All:1:(P)Map:2:usual tests on Map.Make:3:test_add

add 3,4 [3,4; 5,6] == [3,4; 5,6]
```

**Comment by UnixJunkie** [comment]
cc @gasche 
the errors come from testsuite/test_map.ml

**Comment by UnixJunkie** [comment]
does not occur w/ ocaml-5.0.0

**Comment by UnixJunkie** [comment]
could be BatMap.Make(BatInt), BatMap or BatSplay.Map(BatInt)

**Comment by UnixJunkie** [comment]
if my tests are correct, it looks like physical (==) equality is broken for all those maps (while semantic equality (=) is OK):
```
module MBI = BatMap.Make(BatInt)
module BSI = BatSplay.Map(BatInt)
module M = struct
  include BatMap
  let of_list l = of_seq (BatSeq.of_list l)
end;;

(* all OK *)
if MBI.(remove 12 (of_list [3,4; 5,6]) = of_list [3,4; 5,6]) then () else assert(false);;
if M.(  remove 12 (of_list [3,4; 5,6]) = of_list [3,4; 5,6]) then () else assert(false);;
if BSI.(remove 12 (of_list [3,4; 5,6]) = of_list [3,4; 5,6]) then () else assert(false);;

(* all failing *)
if MBI.(remove 12 (of_list [3,4; 5,6]) == of_list [3,4; 5,6]) then () else Printf.eprintf "MBI.remove == KO";;
if M.(  remove 12 (of_list [3,4; 5,6]) == of_list [3,4; 5,6]) then () else Printf.eprintf "  M.remove == KO";;
if BSI.(remove 12 (of_list [3,4; 5,6]) == of_list [3,4; 5,6]) then () else Printf.eprintf "BSI.remove == KO";;
```

**Comment by UnixJunkie** [comment]
the test code I just posted is in fact also failing w/ ocaml-5.0.0; it's possible that the testsuite is just not run by the dune build that we recently switched to

**Comment by UnixJunkie** [comment]
cc @sim642

**Comment by UnixJunkie** [comment]
BatMap.remove 's ocamldoc has a maybe interesting comment:     "before  3.3.0 physical equality was not ensured"

**Comment by sim642** [comment]
The original test failure `All:1:(P)Map:2:usual tests on Map.Make:5:test_remove` is for `Map.Make` here:
https://github.com/ocaml-batteries-team/batteries-included/blob/3e4083c6f742b6db3190ed5874ef0cc1677015fa/testsuite/test_map.ml#L236-L241
If I add `OUnit.assert_failure "reached"` there, then `dune runtest testsuite` does fail, so it is getting executed at least on 5.0.

But looking at `batMap.ml`, it looks like `remove` comes directly from the included `Stdlib.Map.Make`. The standard implementation apparently didn't ensure it before 4.03: https://github.com/ocaml/ocaml/blob/814c063b5433c3030a0475605ea83c93fce3b0a8/stdlib/map.mli#L116.

**Comment by UnixJunkie** [comment]
ok, so those are failing because the stdlib did not support physical equality before ocaml-4.03

**Comment by UnixJunkie** [comment]
I have still this one that fails with ocaml-5.0.0 and that worries me.
```ocaml
let before = BSI.of_list [3,4; 5,6] in assert(before == BSI.remove 12 before);;
```

**Comment by sim642** [comment]
The general documentation for `BatSplay` says:
https://github.com/ocaml-batteries-team/batteries-included/blob/3e4083c6f742b6db3190ed5874ef0cc1677015fa/src/batSplay.mli#L23-L29
So it looks like that it's not even supposed to preserve physical equality.

**Comment by UnixJunkie** [comment]
yes, the documentation for BatSplay.Map.remove is just wrong:
```
val BatSplay.Map.remove key/3 -> 'a t/4 -> 'a t/4                            
    [remove x m] returns a map containing the same bindings as                  
    [m], except for [x] which is unbound in the returned map.                   
    The returned map is physically equal  to the passed one if [x] was          
    already unbound.                                                            
                                                                                
    @before  3.3.0 physical equality was not ensured    
```
It might be wrong for other operations as well that mention physical equality in this module.

**Comment by UnixJunkie** [comment]
I don't even know where the ocamldoc for BatSplay.Map.remove I see in ocp-browser is coming from...

**Comment by sim642** [comment]
I think it's just from the `BatMap` interface and `BatSplay` reuses that.

**Comment by UnixJunkie** [comment]
was fixed by https://github.com/ocaml-batteries-team/batteries-included/pull/1120
